### PR TITLE
[move-lang] Migrate typing/translate::typing_error errors

### PR DESCRIPTION
**Pull request by tnowacki**
- Migrated typing join/subtyping errors (core type system diagnostics)

## Motivation

- The new codespan update gives more IDE friendly error messages 

## Test Plan

- Updated tests

## Related PRs

- Continues work from #8588

### CODE REVIEW

File: language/move-lang/tests/move_check/naming/generics_shadowing_invalid.exp [hunk 1]
```diff
@@ -1,3 +1,14 @@
...
+  ┌─ tests/move_check/naming/generics_shadowing_invalid.move:7:14
+  │
+6 │     fun foo<S>(s1: S, s2: S): S {
+  │                    - The type: 'S'
+7 │         (s1: Self::S);
+  │              ^^^^^^^
+  │              │
+  │              Invalid type annotation
+  │              Is not compatible with: '0x2::M::S'
+
```
Comment (tzakian) [inline language/move-lang/tests/move_check/naming/generics_shadowing_invalid.exp:11]: Not sure if you want to change this in this PR or not, but the ordering of these error messages are a bit funky to me. IMO If I read them from top to bottom, it reads out of order.

I want to read this as "Invalid type annotation" -> "The type: 'S'" -> "Is not compatible with: '0x2::M::S'"

File: language/move-lang/tests/move_check/parser/spec_parsing_inside_fun.exp [hunk 2]
```diff
@@ -7,6 +7,16 @@ error[E04003]: built-in operation not supported
    │         Invalid argument to '+'
    │         Found: '()'. But expected: 'u8', 'u64', 'u128'
 
+error[E04007]: incompatible types
+   ┌─ tests/move_check/parser/spec_parsing_inside_fun.move:32:19
+   │
+32 │         spec {} + 1;
+   │         -------   ^
+   │         │         │
+   │         │         Incompatible arguments to '+'
```
Comment (tzakian) [inline language/move-lang/tests/move_check/parser/spec_parsing_inside_fun.exp:16]: Is there a way to have this first message point to the binary operator (or the whole expression) as opposed to the RHS of the binop?

File: language/move-lang/tests/move_check/parser/spec_parsing_inside_fun.exp [hunk 3]
```diff
@@ -15,6 +25,26 @@ error[E04003]: built-in operation not supported
...
 
+error[E04007]: incompatible types
+   ┌─ tests/move_check/parser/spec_parsing_inside_fun.move:33:9
+   │
+33 │         spec {} && spec {};
+   │         ^^^^^^^
+   │         │
+   │         Incompatible arguments to '&&'
+   │         Is not compatible with: 'bool'
+   │         The type: '()'
```
Comment (tzakian) [inline language/move-lang/tests/move_check/parser/spec_parsing_inside_fun.exp:36]: The order of these is different from the order in the other comment (and out of order).

File: language/move-lang/tests/move_check/translated_ir_tests/move/signer/move_to_missing_signer.exp [hunk 4]
```diff
@@ -1,26 +1,32 @@
...
-   │                   ---------------- Found 1 argument(s) here
+error[E04007]: incompatible types
+  ┌─ tests/move_check/translated_ir_tests/move/signer/move_to_missing_signer.move:4:9
+  │
+4 │         move_to<R>(R { f: false })
+  │         ^^^^^^^^^^^^^^^^^^^^^^^^^^
+  │         │          │
+  │         │          The type: '0x8675309::M::R'
+  │         Invalid call of 'move_to'. Invalid argument for parameter '0'
+  │         Is not compatible with: '&signer'
```
Comment (tzakian) [inline language/move-lang/tests/move_check/translated_ir_tests/move/signer/move_to_missing_signer.exp:9]: Order of these doesn't read quite right in the new format

**Review by tzakian — COMMENTED** [review]
I'm not sure if you want to fix them up in this PR or not, but a a large number of the new error messages read out of order/are a bit funky with the new display style of codespan. Other than the ordering of the messages it looks good to me though. Let me know if you want to fix these up on a separate PR and I'm happy to get this landed first.

File: language/move-lang/tests/move_check/naming/generics_shadowing_invalid.exp (hunk 1 above)
Comment (tnowacki) [reply language/move-lang/tests/move_check/naming/generics_shadowing_invalid.exp]: You should no longer think about these being read from top to bottom, that's why this refactoring has been put off for so long ;) 

With the new codespan reporting, it is designed to be a bit more IDE friendly, as such they are sorted based on the textual location, not based on the order provided. With colors, it comes out a bit cleaner (hopefully).

Here are the errors for this test
image

**Comment by tnowacki** [comment]
I did a bit more of a careful refactoring (the previous was indeed a bit sloppy). I added it as a fixup which should make the comparison between the new and the old a bit more obvious.

Thanks for the suggestions!

File: language/move-lang/tests/move_check/naming/generics_shadowing_invalid.exp (hunk 1 above)
Comment (tzakian) [reply language/move-lang/tests/move_check/naming/generics_shadowing_invalid.exp]: Makes a lot more sense with colors!

**Review by tzakian — APPROVED** [review]
LGTM. The error messages make a lot more sense in color!

**Comment by tnowacki** [comment]
>  The error messages make a lot more sense in color
I do deeply worry about this for either colorblind folks or for different terminal setups (or just even for our own tests).
It would be interesting if there was something we could do for those modes

File: language/move-lang/tests/move_check/parser/spec_parsing_inside_fun.exp (hunk 2 above)
Comment (tnowacki) [reply language/move-lang/tests/move_check/parser/spec_parsing_inside_fun.exp:16]: Didn't see this one, let me fix that before landing

**Comment by tzakian** [comment]
Very good point about how this would work for colorblind folks. 

I had a look, and it appears as though codespan already has some support for different display styles [link](https://docs.rs/codespan-reporting/0.11.1/codespan_reporting/term/enum.DisplayStyle.html), so it might be worthwhile for us to see if they have any plans on supporting a`NoColor` display style or similar (or we might want to make an upstream contribution to it if not).

**Comment by tnowacki** [comment]
> Very good point about how this would work for colorblind folks.
> 
> I had a look, and it appears as though codespan already has some support for different display styles [link](https://docs.rs/codespan-reporting/0.11.1/codespan_reporting/term/enum.DisplayStyle.html), so it might be worthwhile for us to see if they have any plans on supporting a`NoColor` display style or similar.

The crate is largely trying to copy what Rust is doing for errors. I would be curious what Rust does for this sort of problem. Might be a easy way to agree on a design with the codespan folks

File: language/move-lang/tests/move_check/parser/spec_parsing_inside_fun.exp [hunk 5]
```diff
@@ -7,6 +7,15 @@ error[E04003]: built-in operation not supported
...
    │         Found: '()'. But expected: 'u8', 'u64', 'u128'
 
+error[E04007]: incompatible types
+   ┌─ tests/move_check/parser/spec_parsing_inside_fun.move:32:17
+   │
+32 │         spec {} + 1;
+   │         ------- ^ - Found: integer. It is not compatible with the other type.
+   │         │       │  
+   │         │       Incompatible arguments to '+'
+   │         Found: '()'. It is not compatible with the other type.
```
Comment (tzakian) [inline language/move-lang/tests/move_check/parser/spec_parsing_inside_fun.exp:17]: ❤️

**Review by tzakian — DISMISSED** [review]
Looks great! Loving the new and improved messages (especially around binops!)

**Review by tzakian — APPROVED** [review]

**Pull request closed without merging**